File: Laboratori/Ghigo/Lab0/Produttore.java
import java.io.BufferedReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.UnsupportedEncodingException;

// Proposta, rendere produttore un filtro

public class Produttore {
	public static void main(String[] args) {		
		BufferedReader in = null; // dichiarazione di variabile in di tipo BufferedReader

		if (args.length != 1){ // se la lunghezza dell'array args è diversa da 1
			System.out.println("Utilizzo: produttore <inputFilename>");
			System.exit(0);
		}

		System.out.println("Inserisci la nuova riga, EOF per terminare");

		FileWriter fout;
		String inputl = null;
		
		try {
			fout = new FileWriter(args[0]); // inizializzazione di fout con un nuovo FileWriter che ha come parametro args[0]
            in = new BufferedReader(new InputStreamReader(System.in)); // inizializzazione di in con un nuovo BufferedReader che ha come parametro un nuovo InputStreamReader che ha come parametro System.in	
			while((inputl = in.readLine()) != null){ //till EOF (CTRL+D)
				fout.write(inputl+"\n", 0, inputl.length()+1);
			}	
			fout.close();
		}
	    catch (IOException e) { 
			e.printStackTrace();
			System.exit(2); // uscita con errore, intero positivo a livello di sistema Unix
		}
	}
}

// done

File: Laboratori/Ghigo/Lab0/produttore.c
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#define MAX_STRING_LENGTH 256

// proposta, rendere il produttore un filtro
int main(int argc, char *argv[]) {
    int   fd, readValues, bytes_to_write, written;   // variabili per la gestione del file
    char *file_out; // nome del file
    char  riga[MAX_STRING_LENGTH], buf[MAX_STRING_LENGTH]; // buffer per la riga
    char  read_char; // variabile per la lettura di un carattere
    int   nread;    // numero di byte letti

    // controllo numero argomenti
    if (argc != 2) {
        perror(" numero di argomenti sbagliato\n 1: nome del file\n");
        exit(1);
    }

    file_out = argv[1]; // nome del file
    

    fd = open(file_out, O_WRONLY | O_CREAT | O_TRUNC, 00640);
    if (fd < 0) {
        perror("P0: Impossibile creare/aprire il file");
        exit(2);
    }

    printf("Inserisci la nuova riga, ^D per terminare\n");

     while (gets(riga)!=NULL){ // Fino ad EOF
        printf("Inserisci la nuova riga, ^D per terminare\n");
        /* la gets legge tutta la riga, separatori inclusi, e trasforma il fine
           linea in fine stringa */
        riga[strlen(riga) + 1] = '\0'; // aggiungo 0 binario per permettere una corretta strlen()
        riga[strlen(riga)] = '\n'; // aggiungo il fine linea
        written = write(fd, riga, strlen(riga)); // uso della primitiva
        if (written < 0) {
            perror("P0: errore nella scrittura sul file");
            exit(3);
        }
    }
    close(fd);
}

// done